### PR TITLE
Fix various bugs with LoRA Dreambooth and Dreambooth script

### PR DESCRIPTION
**Pull request by patrickvonplaten**
This PR fixes the the issues #3279, #3346, #3363 and #3296. Here we also make sure that LoRA checkpointing is in a nice format. After this PR the format of LoRA checkpointing for dreambooth has the format:

```
├── optimizer.bin
├── pytorch_lora_weights.bin
├── random_states_0.pkl
├── scaler.pt
└── scheduler.bin
```

which allows one to easily load lora weights from checkpoints.

Script is tested for resuming from checkpointing for both single and multi GPU setups.

### CODE REVIEW

File: examples/dreambooth/train_dreambooth.py [hunk 1]
```diff
@@ -733,36 +733,34 @@ def main(args):
         args.pretrained_model_name_or_path, subfolder="unet", revision=args.revision
     )
 
-    # `accelerate` 0.16.0 will have better support for customized saving
-    if version.parse(accelerate.__version__) >= version.parse("0.16.0"):
```
Comment (patrickvonplaten) [inline examples/dreambooth/train_dreambooth.py:737]: min version of accelerate is already 0.16 => let's remove this check here

File: src/diffusers/loaders.py [hunk 2]
```diff
@@ -1059,7 +1085,7 @@ def _load_text_encoder_attn_procs(
     def save_lora_weights(
         self,
         save_directory: Union[str, os.PathLike],
-        unet_lora_layers: Dict[str, torch.nn.Module] = None,
+        unet_lora_layers: Dict[str, Union[torch.nn.Module, torch.Tensor]] = None,
```
Comment (patrickvonplaten) [inline src/diffusers/loaders.py:1088]: Let's allow to also pass the lora layers as low level torch.Tensors state dicts

File: src/diffusers/loaders.py [hunk 3]
```diff
@@ -77,10 +80,17 @@ def map_to(module, state_dict, *args, **kwargs):
...
+                if k in key:
+                    return key.split(k)[0] + k
+
+            raise ValueError(f"There seems to be a problem with the state_dict: {set(state_dict.keys())}. {key} has to have one of {self.split_keys}.")
+
         def map_from(module, state_dict, *args, **kwargs):
             all_keys = list(state_dict.keys())
             for key in all_keys:
-                replace_key = key.split(".processor")[0] + ".processor"
+                replace_key = remap_key(key, state_dict)
```
Comment (patrickvonplaten) [inline src/diffusers/loaders.py:93]: We need to be able to split this also according to other keys

File: examples/dreambooth/train_dreambooth_lora.py [hunk 4]
```diff
@@ -743,9 +743,53 @@ def main(args):
         )
         temp_pipeline._modify_text_encoder(text_lora_attn_procs)
         text_encoder = temp_pipeline.text_encoder
-        accelerator.register_for_checkpointing(text_encoder_lora_layers)
         del temp_pipeline
 
+    # create custom saving & loading hooks so that `accelerator.save_state(...)` serializes in a nice format
+    def save_model_hook(models, weights, output_dir):
```
Comment (patrickvonplaten) [inline examples/dreambooth/train_dreambooth_lora.py:749]: Let's make sure the lora layers are saved in a more user-friendly way here

File: examples/dreambooth/train_dreambooth_lora.py [hunk 5]
```diff
@@ -724,7 +724,6 @@ def main(args):
 
     unet.set_attn_processor(unet_lora_attn_procs)
     unet_lora_layers = AttnProcsLayers(unet.attn_processors)
-    accelerator.register_for_checkpointing(unet_lora_layers)
```
Comment (patrickvonplaten) [inline examples/dreambooth/train_dreambooth_lora.py:727]: We now save LoRA layers in better format

File: src/diffusers/loaders.py [hunk 6]
```diff
@@ -843,10 +855,13 @@ def load_lora_weights(self, pretrained_model_name_or_path_or_dict: Union[str, Di
             text_encoder_lora_state_dict = {
                 k.replace(f"{self.text_encoder_name}.", ""): v for k, v in state_dict.items() if k in text_encoder_keys
             }
+            attn_procs_text_encoder = self._load_text_encoder_attn_procs(text_encoder_lora_state_dict)
             if len(text_encoder_lora_state_dict) > 0:
```
Comment (patrickvonplaten) [inline src/diffusers/loaders.py:859]:
```suggestion
            if len(text_encoder_lora_state_dict) > 0:
                attn_procs_text_encoder = self._load_text_encoder_attn_procs(text_encoder_lora_state_dict)
```

File: src/diffusers/loaders.py [hunk 7]
```diff
@@ -880,6 +901,13 @@ def new_forward(x):
                 # Monkey-patch.
                 module.forward = new_forward
 
+    @property
+    def text_encoder_lora_procs(self):
+        for name, _ in self.text_encoder.named_modules():
+            if any(x in name for x in TEXT_ENCODER_TARGET_MODULES):
+                # Retrieve the module and its corresponding LoRA processor.
+                self.text_encoder.get_submodule(name)
+
```
Comment (patrickvonplaten) [inline src/diffusers/loaders.py:910]:
```suggestion
```

File: src/diffusers/loaders.py [hunk 8]
```diff
@@ -879,7 +900,6 @@ def new_forward(x):
 
                 # Monkey-patch.
                 module.forward = new_forward
```
Comment (patrickvonplaten) [inline src/diffusers/loaders.py:902]:
```suggestion
                module.forward = new_forward
                
```

File: src/diffusers/loaders.py [hunk 9]
```diff
@@ -879,7 +900,7 @@ def new_forward(x):
...
-
+
```
Comment (patrickvonplaten) [inline src/diffusers/loaders.py:903]:
```suggestion

```

File: examples/dreambooth/train_dreambooth_lora.py [hunk 10]
```diff
@@ -743,9 +742,63 @@ def main(args):
...
+        # there are only two options here. Either are just the unet attn processor layers
+        # or there are the unet and text encoder atten layers
+        unet_lora_layers_to_save = None
+        text_encoder_lora_layers_to_save = None
+
+        for model in models:
+            state_dict = model.state_dict()
+
+            text_encoder_keys = accelerator.unwrap_model(text_encoder_lora_layers).state_dict().keys()
+            unet_keys = accelerator.unwrap_model(unet_lora_layers).state_dict().keys()
```
Comment (sayakpaul) [inline examples/dreambooth/train_dreambooth_lora.py:758]: This can go before the loop, no?

File: examples/dreambooth/train_dreambooth_lora.py [hunk 11]
```diff
@@ -743,9 +742,63 @@ def main(args):
...
+        # so that the we don't accidentally override the LoRA layers of
+        # unet_lora_layers and text_encoder_lora_layers which are stored in `models`
+        # with new torch.nn.Modules / weights. We simply use the pipeline class as 
+        # an easy way to load the lora checkpoints
+        temp_pipeline = DiffusionPipeline.from_pretrained(
+            args.pretrained_model_name_or_path,
+            revision=args.revision,
+            torch_dtype=weight_dtype,
+        )
+        temp_pipeline.load_lora_weights(input_dir)
```
Comment (sayakpaul) [inline examples/dreambooth/train_dreambooth_lora.py:787]: Clean and tight 🎸

File: examples/dreambooth/train_dreambooth_lora.py [hunk 12]
```diff
@@ -743,9 +742,63 @@ def main(args):
...
+            unet_lora_layers=unet_lora_layers_to_save,
+            text_encoder_lora_layers=text_encoder_lora_layers_to_save,
+        )
+
+    def load_model_hook(models, input_dir):
+        # Note we DON'T pass the unet and text encoder here an purpose 
+        # so that the we don't accidentally override the LoRA layers of
+        # unet_lora_layers and text_encoder_lora_layers which are stored in `models`
+        # with new torch.nn.Modules / weights. We simply use the pipeline class as 
+        # an easy way to load the lora checkpoints
```
Comment (sayakpaul) [inline examples/dreambooth/train_dreambooth_lora.py:781]: I don't fully understand why the override will happen, though.

File: examples/dreambooth/train_dreambooth_lora.py [hunk 13]
```diff
@@ -743,9 +742,63 @@ def main(args):
...
+        temp_pipeline = DiffusionPipeline.from_pretrained(
+            args.pretrained_model_name_or_path,
+            revision=args.revision,
+            torch_dtype=weight_dtype,
+        )
+        temp_pipeline.load_lora_weights(input_dir)
+
+        # load lora weights into models
+        models[0].load_state_dict(AttnProcsLayers(unet.attn_processors).state_dict())
```
Comment (sayakpaul) [inline examples/dreambooth/train_dreambooth_lora.py:790]: Shouldn't the UNet be accessed from the temporary pipeline i.e., `temp_pipeline.unet`?

**Review by sayakpaul — APPROVED** [review]
Thanks a lot for this. This has been a somewhat longstanding issue. Glad that we now have a decent resolution.

File: examples/dreambooth/train_dreambooth_lora.py (hunk 13 above)
Comment (pcuenca) [reply examples/dreambooth/train_dreambooth_lora.py:790]: Good catch, I agree!

Comment (patrickvonplaten) [reply examples/dreambooth/train_dreambooth_lora.py:791]: Ufff yes good catch! This would have led to some ugly silent errors

**Comment by patrickvonplaten** [comment]
Cool this is good to go! Thanks a mille for the in-depth reviews @pcuenca @sayakpaul !

**Pull request merged**